File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int goodNodes(TreeNode root) {
        return helper(root.left, root.val) + helper(root.right, root.val) + 1;
    }

    public int helper(TreeNode root, int maxValue) {
        if (root == null) return 0;
        int goodNode = 0;
        
        if (root.val >= maxValue) {
            goodNode += 1;
            maxValue = root.val;
        }

        int leftGoodNodes = helper(root.left, maxValue);
        int rightGoodNodes = helper(root.right, maxValue);

        return leftGoodNodes + rightGoodNodes + goodNode;
    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if root is None:
            return True
        
        return self.helper(root)[0]
        
    def helper(self, root: Optional[TreeNode]) -> Tuple[bool, int]:
        if root is None:
            return (True, 0)
        
        lb, lh = self.helper(root.left)
        rb, rh = self.helper(root.right)

        if not lb or not rb:
            return (False, -1)
        
        if abs(lh - rh) > 1:
            return (False, -1)
        
        return (True, max(lh, rh) + 1)
